contrib/repro_test.bzl

```diff
@@ -0,0 +1,222 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test rule for verifying a given container_image target produces a
+   deterministic Docker image.
+
+This rule builds the given container_image twice in isolation to determine if
+the given image is deterministic.
+It does so by mounting the provided Bazel project files into the 'base' image.
+Next it starts two containers and builds the supplied container_image target
+(being 'image') inside each of the containers. The rule then extracts the
+output files for the test image. Finally, both extracted images are compared.
+A test image is considered to be reproducible when both instances of the same
+image have the same digest and ID. On a mismatch of one of those values, the
+test fails and the container_diff tool
+(https://github.com/GoogleContainerTools/container-diff#container-diff)
+produces the summary of their differences.
+
+Args:
+    image: A container_image target relative to the project's root to test
+           for determinism.
+    base: An image target to build and reproduce the 'image 'in.
+          This image must have Bazel and docker installed and available in the PATH.
+    workspace_file: The WORKSPACE file of the project containing the 'image'
+                    target to help detect project's root path.
+    container_diff_args: (optional) Args to the container_diff tool as specified here:
+                         https://github.com/GoogleContainerTools/container-diff#quickstart
+
+Example:
+
+container_repro_test(
+    name = "set_cmd_repro_test",
+    image = "//tests/container:set_cmd",
+    base = "@bazel_latest//image",
+    workspace_file = "//:WORKSPACE",
```

Comment:
you can probably make this a private attr if the expectation is that it will always receive "//:WORKSPACE" as value.

Reply:
Yes, the expectation is that it will always receive the "//:WORKSPACE" value, but external repos using the `container_repro_test` target will need to pass their own WORKSPACE files and making this private will always point at the rules_docker project WORKSPACE file.

Reply:
sgtm.